### ISSUE TITLE
Running several 'bosh prepare' launched from distinct CWDs crash and leave files behind

### ISSUE THREAD

**Issue by prioux** [boutiques/boutiques#404]
This bug report applies to the 'develop' branch, currently at 9d4cd352e70aaed8ea2209c268c65ec443380e57 .

Given a Boutiques descriptors that requires a Singularity container, if I try to:

* run several 'bosh prepare' in parallel
* where each bosh prepare is invoked from a distinct current working directory
* where the target directory --imagepath for the image is the same for all of them

then:

* while the `bosh prepare` are running, several image file will be created an pulled in parallel in the target directory (with distinct temp extensions)
* then all `bosh prepare` will crash with a python stack trace
* the target directory for the pull will finally contain a single spurious `.tmp` copy of the image file

Here is a shell session that demonstrate the problem:

```
unix> ls
boutiques_adjusted.json
unix> mkdir -p t/cwd_a t/cwd_b t/pull
unix> cd t/cwd_a
unix> bosh exec prepare --imagepath ../pull ../../boutiques_adjusted.json &
[1] 28466
unix> ls -l
total 12
drwxr-xr-x 2 prioux prioux 4096 Feb 10 13:53 shots47s-bids-fmriprep-1.2.3.simg-lock/
unix> cd ../cwd_b
unix> bosh exec prepare --imagepath ../pull ../../boutiques_adjusted.json &
[2] 28596
unix> ls -l
total 12
drwxr-xr-x 2 prioux prioux 4096 Feb 10 13:54 shots47s-bids-fmriprep-1.2.3.simg-lock/
unix> ls -l ../pull
total 1720332
-rw-r--r-- 1 prioux prioux 1525678080 Feb 10 13:54 shots47s-bids-fmriprep-1.2.3.simg.tmp.XIKbdB
-rw-r--r-- 1 prioux prioux  235929600 Feb 10 13:54 shots47s-bids-fmriprep-1.2.3.simg.tmp.vKYr_D
(time passes...)
unix> Traceback (most recent call last):
  File "/home/users/prioux/.local/bin/bosh", line 9, in <module>
    load_entry_point('boutiques==0.5.18', 'console_scripts', 'bosh')()
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/bosh.py", line 523, in bosh
    out = execute(*params)
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/bosh.py", line 211, in execute
    container_location = executor.prepare()[1]
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/localExec.py", line 436, in prepare
    return self._pullSingImage(conName, conIndex, conImage)
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/localExec.py", line 483, in _pullSingImage
    os.rename(conNameTmp, conName)
OSError: [Errno 2] No such file or directory
[2]+  Exit 1                  bosh exec prepare --imagepath ../pull ../../boutiques_adjusted.json
(time passes...)
unix>
unix> Traceback (most recent call last):
  File "/home/users/prioux/.local/bin/bosh", line 9, in <module>
    load_entry_point('boutiques==0.5.18', 'console_scripts', 'bosh')()
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/bosh.py", line 523, in bosh
    out = execute(*params)
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/bosh.py", line 211, in execute
    container_location = executor.prepare()[1]
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/localExec.py", line 436, in prepare
    return self._pullSingImage(conName, conIndex, conImage)
  File "/home/users/prioux/src/boutiques/tools/python/boutiques/localExec.py", line 483, in _pullSingImage
    os.rename(conNameTmp, conName)
OSError: [Errno 2] No such file or directory
[1]+  Exit 1                  bosh exec prepare --imagepath ../pull ../../boutiques_adjusted.json
unix> ls -l ../pull
total 4332364
-rwxr-xr-x 1 prioux prioux 4436332575 Feb 10 13:56 shots47s-bids-fmriprep-1.2.3.simg.tmp*
```

**Comment by prioux** [comment]
The lock directory that are created should be created in the same place as where the image is pulled.

**Comment by erinb90** [comment]
When I implemented it I had assumed that `bosh prepare` would be invoked from the same working directory in all processes. It should be straightforward to change the location of the lockdir from CWD to the image location.

**Comment by prioux** [comment]
Ideally, the lock directory should be whatever path is provided with `--imagepath`, appended with the `-lock`. If there is no `--imagepath`, use the same rule with the internal value for the image. The point is the lock directory should be just besides where the image will be.

**Comment by prioux** [comment]
I made some tests and it seems to work fine now @erinb90  !

I'm not sure if I'm the one supposed to close this issue now, @glatard , but I'll do it this time. Let me know if it was supposed to be you or Erin.

**Comment by prioux** [comment]
I'm re-opening this issue, found a slight limitation.

The prepare command fails when the imagepath given is just a plain basename, e.g. in `--imagepath abc.simg`

**Comment by prioux** [comment]
The error is here:

```
  File "/blahblah/boutiques/tools/python/boutiques/localExec.py", line 468, in _singConExists
    return conName in os.listdir(imageDir)
```

the reason being that `imageDir` is an empty string.

**Comment by prioux** [comment]
Fixed in #413 (thanks Erin). We can close the issue.
